### PR TITLE
Support Data Streams in OpenSearch

### PR DESCRIPTION
**Pull request by ketanv3**
### Description

This PR aims to add the missing `MetadataFieldMapper` required to parse the `_data_stream_timestamp` metadata field mapping used to create data streams and to perform timestamp field validation on the ingested documents. This PR also allows users to change the default timestamp field name (from "@timestamp") as required using an index template.
 
### Issues Resolved
+ Closes https://github.com/opensearch-project/OpenSearch/issues/675
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Signed-off-by: Ketan Verma <ketan9495@gmail.com>

### CODE REVIEW

**Comment by nknize** [comment]
start gradle check

File: server/src/main/java/org/opensearch/cluster/metadata/ComposableIndexTemplate.java [hunk 1]
```diff
@@ -264,51 +266,74 @@ public String toString() {
...
+                        "enabled", true,
+                        "timestamp_field", getTimestampField().toMap()
+                    ))
+                )
+            );
         }
 
         @Override
         public void writeTo(StreamOutput out) throws IOException {
+            out.writeOptionalWriteable(timestampField);
```
Comment (nknize) [inline server/src/main/java/org/opensearch/cluster/metadata/ComposableIndexTemplate.java:315]: Also be sure to wrap the companion writeTo in a version check.
```suggestion
            if (out.getVersion().onOrAfter(Version.V1_0_0)) {
                out.writeOptionalWriteable(timestampField);
            }    
```

File: server/src/main/java/org/opensearch/cluster/metadata/ComposableIndexTemplate.java [hunk 2]
```diff
@@ -264,51 +266,74 @@ public String toString() {
...
+        }
+
+        public DataStreamTemplate(TimestampField timestampField) {
+            this.timestampField = timestampField;
         }
 
-        public String getTimestampField() {
-            return FIXED_TIMESTAMP_FIELD;
+        public DataStreamTemplate(StreamInput in) throws IOException {
+            this.timestampField = in.readOptionalWriteable(TimestampField::new);
```
Comment (nknize) [inline server/src/main/java/org/opensearch/cluster/metadata/ComposableIndexTemplate.java:292]: This is a new (optional) field in OpenSearch 1.0.0 for a class that existed in previous versions. To support backwards compatibility we need to wrap in a version check.
```suggestion
            if (in.getVersion().onOrAfter(Version.V_1_0_0)) {
                this.timestampField = in.readOptionalWriteable(TimestampField::new);
            }
```

**Review by nknize — CHANGES_REQUESTED** [review]
Looking good; one backwards compatibility addition needed

**Review by matznermatzner — APPROVED** [review]

File: server/src/main/java/org/opensearch/cluster/metadata/ComposableIndexTemplate.java (hunk 1 above)
Comment (ketanv3) [reply server/src/main/java/org/opensearch/cluster/metadata/ComposableIndexTemplate.java]: Thanks for pointing this out. I have added this check along with UTs.

File: server/src/main/java/org/opensearch/cluster/metadata/ComposableIndexTemplate.java (hunk 2 above)
Comment (ketanv3) [reply server/src/main/java/org/opensearch/cluster/metadata/ComposableIndexTemplate.java]: Done.

**Comment by nknize** [comment]
start gradle check

**Review by nknize — APPROVED** [review]
LGTM! Nice to get this out there!

**Comment by nknize** [comment]
Thank you for working this enhancement. Can you open a backport PR to the 1.x branch? The pending label will stay applied until we merge to 1.x.

**Comment by ketanv3** [comment]
@nknize Created this backport PR – https://github.com/opensearch-project/OpenSearch/pull/713

**Pull request merged**